src/mds/Server.cc

```diff
@@ -200,10 +200,10 @@ void Server::dispatch(Message *m)
   }
 
   // active?
-  if (!mds->is_active()) {
-    if (m->get_type() == CEPH_MSG_CLIENT_REQUEST &&
-	(mds->is_reconnect() || mds->get_want_state() == CEPH_MDS_STATE_RECONNECT)) {
-      MClientRequest *req = static_cast<MClientRequest*>(m);
+  // handle_slave_request()/handle_client_session() will wait if necessary
+  if (m->get_type() == CEPH_MSG_CLIENT_REQUEST && !mds->is_active()) {
```

Comment:
this is great, you are trying to clean-up the massive code in Server::dispatch, below is my minor suggestion:
 maybe move this big "if condition" to handle_client_request would make it more clean, since it's only about CEPH_MSG_CLIENT_REQUEST now.  and put them in a function which only do something when mds is not active.